### ISSUE TITLE
Possible regression between 1.2.162.1 and 1.2.170.0/1.2.176.1 when rendering using VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT

### ISSUE THREAD

**Issue by mbechard** [KhronosGroup/Vulkan-ValidationLayers#2851]
**Describe the Issue**
I'm rendering to a single slice of a 3D texture that has been created using VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT. The subresource for the view is only referencing a single layer, and it is a 2D view. The renderpass has SHADER_READ_OPTIMAL as both the initial and final layout for the attachment.
This runs without errors in SDK 1.2.162.1. Using 1.2.170.0 and 1.2.176.1, I get the below error for all of the layers except for the one I'm trying to render to. In this case, I'm rendering to layer 7.
Unless I'm missing something, it would seem the validation layers are failing to keep track that the render pass has transitioned all the layers to COLOR_ATTACHMENT_OPTIMAL.

**Valid Usage ID**
`Validation Error: [ UNASSIGNED-CoreValidation-DrawState-InvalidImageLayout ] Object 0: handle = 0x2b5626fb5e0, name = uniqueID: 3, Main Content, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x4dae5635 | Submitted command buffer expects VkImage 0xd3f98f0000000b70[uniqueID: 2311] (subresource: aspectMask 0x1 array layer 0, mip level 0) to be in layout VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL--instead, current layout is VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL.`

**Environment:**
 - OS: Windows 10
 - GPU: AMD 5500 XT
 - SDK or header version if building from repo: 1.2.176.1
 - Options enabled (synchronization, best practices, etc.):

**Additional context**

**Comment by sfricke-samsung** [comment]
So just doing a quick diff of `ValidateCmdBufImageLayouts` (function that throws the validation error) between [1.2.162](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/7cf6e361ee511723ae58fc17ffe88f16965be42b/layers/buffer_validation.cpp#L4144) and ToT seems the only real difference is

```patch
- const auto &initial_layout_map = subres_map->GetInitialLayoutMap();
+ const auto &layout_map = subres_map->GetLayoutMap();
//...
- auto pos = initial_layout_map.begin();
- const auto end = initial_layout_map.end();
- sparse_container::parallel_iterator<const ImageSubresourceLayoutMap::LayoutMap> current_layout(*overlay_map, *global_map, pos->first.begin);
+ auto pos = layout_map.begin();
+ const auto end = layout_map.end();
+ sparse_container::parallel_iterator<const GlobalImageLayoutRangeMap> current_layout(*overlay_map, *global_map, pos->first.begin);
while (pos != end) {
-     VkImageLayout initial_layout = pos->second;
+     VkImageLayout initial_layout = pos->second.initial_layout;
```

Haven't fully parsed everything yet, but dropping as some potential useful info

**Comment by jeremyg-lunarg** [comment]
@sfricke-samsung, there's been a ton of performance work on image layouts in the past few months and that code diff is just the tip of the iceberg. I'll take a look.

**Comment by jeremyg-lunarg** [comment]
@mbechard https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2352/ looks like it is related to VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT validation, and it came in during 1.2.170.   It is possible that your problem is either caused by this change OR that the change is incomplete and similar logic needs to be changed in other parts of validaiton. (or I could be completely wrong and the culprit is a different change...)

Is it possible for you to provide a partial code snippet, api dump or replay file to help me reproduce the problem here? Thanks!

**Comment by mbechard** [comment]
Sure, here is an API dump. The first framebuffer created to render into the 3D texture is done at line 1,838,429. Let me know if this isn't helpful. Sorry the file is so large.

[api.zip](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/6586525/api.zip)

**Comment by jeremyg-lunarg** [comment]
Thanks I think I can work with this. Looking at the full history of the image that got the error will be possible in this dump and that's what I need to start with.

**Comment by jeremyg-lunarg** [comment]
I was able to write a test case to reproduce this problem.  Working on a fix.

**Comment by jeremyg-lunarg** [comment]
After looking into this further, I think validation is working as it should. There's a specific statement about this case in the [renderpass chapter](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#renderpass) of the vulkan specification:

> Automatic layout transitions apply to the entire image subresource attached to the framebuffer. If the attachment view is a 2D or 2D array view of a 3D image, even if the attachment view only refers to a subset of the slices of the selected mip level of the 3D image, automatic layout transitions apply to the entire subresource referenced which is the entire mip level in this case.

These 2 changes to CTS might also be of interest:
https://github.com/KhronosGroup/VK-GL-CTS/commit/3b7a9eb53d78e3d0fe7967a5560b57e62d167291
https://github.com/KhronosGroup/VK-GL-CTS/commit/d03d91bf04264b88a36a58373a2ca1fe7fb29348

I'm adding a [test case](https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2868) to cover this scenario, since we don't currently have a ton of coverage for this type of image.

**Comment by mbechard** [comment]
Right, the 'check' itself is correct. The issue is the renderpass has caused an automatic transition of all of those layers, but the validations code isn't tracking those changes, and thus the check is failing.

**Comment by jeremyg-lunarg** [comment]
With a fixed up test case and a bunch of bisecting I was able to reproduce the problem and then find that it was fixed by 884579ec2928a26cffac0573101f48efe82d87de, which merged shortly after the 1.2.176.1 SDK.

Oddly, 884579ec2928a26cffac0573101f48efe82d87de was a fix for 5363130b75709e96e75cfecb75750c7ffe458b3e which is also NOT in the 1.2.176.1 SDK.  I suspect that the changes in #2352 exposed this problem by fixing the handling of VK_IMAGE_CREATE_2D_ARRAY_COMPATIBLE_BIT ImageViews and these other changes got lucky and fixed something we didn't know was broken.  I'm going to merge the new test case to make sure we don't need quite as much luck next time.
